### PR TITLE
Refactor: Implement read and write dmk with a UnitTest

### PR DESCRIPTION
**Pull request by WHUweiqingzhou**
Implement #2801

Change List:
1. add a new subdirectory `module_elecstate/module_dm/` to store density matrix source files
2. In `module_elecstate/module_dm/`, implement two new functions to read and write dmk `read_dmk()` and `write_dmk()`
3. add corresponding UnitTest of `read_dmk()` and `write_dmk()`

### CODE REVIEW

File: source/module_elecstate/module_dm/dmk_io.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
+#include "dmk_io.h"
+#include "module_base/parallel_common.h"
+#include "module_base/timer.h"
+
+// output DM_k
+void ModuleIO::write_dmk(
```
Comment (hongriTianqi) [inline source/module_elecstate/module_dm/dmk_io.cpp]: I suggest to change the namespace from ModuleIO to elecstate

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_elecstate/module_dm/test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+remove_definitions(-D__DEEPKS)
+remove_definitions(-D__CUDA)
+remove_definitions(-D__ROCM)
+remove_definitions(-D__MPI)
+
+install(DIRECTORY support DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+
+AddTest(
+  TARGET read_dm_gamma
```
Comment (hongriTianqi) [inline source/module_elecstate/module_dm/test/CMakeLists.txt]: Please rename the test

**Review by hongriTianqi — APPROVED** [review]

File: source/module_elecstate/module_dm/test/test_readdmgamma.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+			}
+			delete[] DM[is];
+			delete[] DM_R[is];
+		}
+		delete[] DM;
+		delete[] DM_R;
+	}
+};
+
+TEST_F(DMIOTest,Read)
```
Comment (hongriTianqi) [inline source/module_elecstate/module_dm/test/test_readdmgamma.cpp]: This is a repeat of unit test in module_io, and I suggest not to add it here.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_elecstate/module_dm/dmk_io.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#define DMK_IO_H
+
+#include <string>
+#include "module_cell/klist.h"
+#include "module_base/complexmatrix.h"
+#include "module_base/matrix.h"
+
+namespace elecstate
+{
+	void write_dmk(const K_Vectors& kv,const int& ik,const std::string &fn, const int &precision,std::vector<ModuleBase::ComplexMatrix> &dm_k);
```
Comment (hongriTianqi) [inline source/module_elecstate/module_dm/dmk_io.h]: Please add comments on header file

File: source/module_elecstate/module_dm/dmk_io.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        //ofs << "\n " << ef << " (fermi energy)";
+
+        ofs << "\n  " << GlobalV::NLOCAL << " " << GlobalV::NLOCAL << std::endl;
+
+        ofs << std::setprecision(precision);
+        ofs << std::scientific;
+
+    }
+
+    for(int i=0; i<GlobalV::NLOCAL; ++i)
```
Comment (hongriTianqi) [inline source/module_elecstate/module_dm/dmk_io.cpp]: GlobalV::NLOCAL is suggested to be removed from beginning.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_elecstate/module_dm/dmk_io.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    const int &precision,
+    std::vector<ModuleBase::ComplexMatrix> &dm_k)
+{
+    ModuleBase::TITLE("elecstate","write_dmk");
+    ModuleBase::timer::tick("elecstate","write_dmk");
+
+    time_t start, end;
+    std::ofstream ofs;
+
+    if(GlobalV::MY_RANK==0)
```
Comment (hongriTianqi) [inline source/module_elecstate/module_dm/dmk_io.cpp]: GlobalV::MY_RANK is not set in the unit test, which might cause errors.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

**Review by dyzheng — APPROVED** [review]

File: source/module_elecstate/module_dm/test/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+remove_definitions(-D__ROCM)
+remove_definitions(-D__MPI)
+
+install(DIRECTORY support DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+
+AddTest(
+  TARGET dm_k_io
+  LIBS ${math_libs} base device cell_info symmetry
+  SOURCES test_dmk_io.cpp ../dmk_io.cpp ${ABACUS_SOURCE_DIR}/module_io/output.cpp
+  ${ABACUS_SOURCE_DIR}/module_cell/klist.cpp
```
Comment (dyzheng) [inline source/module_elecstate/module_dm/test/CMakeLists.txt]: In my opinion, add a mock of klist.h for functions you need is a better solution.

**Review by dyzheng — APPROVED** [review]

**Review by hongriTianqi — APPROVED** [review]

**Pull request merged**